### PR TITLE
API - Handle blocks without blobs gracefully

### PR DESCRIPTION
**Pull request by qu0b**
## Issue Addressed

On my local testnet after the deneb fork my lighthouse node returns for a block that does exist.

```json
  {
    "code": 404,
    "message": "NOT_FOUND: No blobs with block root 0x3956…d78c found in the store",
    "stacktraces": []
  }
  ```
I would expect the endpoint to handle blocks without blobs gracefully and would return an empty array. The beaconchain API specification also defines the 404 error code as an error where the [block was not found](https://github.com/ethereum/beacon-APIs/blob/master/apis/beacon/blob_sidecars/blob_sidecars.yaml#L49).

## Proposed Changes

This PR proposes to handle requests gracefully if there are no blobs for a block.

## Additional Info

```bash
BLOCK_ROOT=0x3956e3e21be49b23abbfbb7c05c00aa078bc03be4f37dc1590b036bafd10d78c
$ curl localhost:5000/eth/v1/beacon/blob_sidecars/$BLOCK_ROOT
{
    "code": 404,
    "message": "NOT_FOUND: No blobs with block root 0x3956…d78c found in the store",
    "stacktraces": []
  }
$ curl localhost:5000/eth/v1/beacon/headers/$BLOCK_ROOT
  {
    "execution_optimistic": false,
    "finalized": false,
    "data": {
      "root": "0x3956e3e21be49b23abbfbb7c05c00aa078bc03be4f37dc1590b036bafd10d78c",
      "canonical": true,
      "header": {
        "message": {
          "slot": "130",
          "proposer_index": "45",
          "parent_root": "0xf550895fcd6b5d6a78c00b2c2fe0bd32d878031d5562b9b9ad8183c431fbe88d",
          "state_root": "0x249cc0a46ff183f990cd706d0167677b8fa2d99fad270e9d7361575c226846f6",
          "body_root": "0x76bc18d7cc0d0ac32219633b1f05d67c421bac15f59511a924c705c2e8cb128f"
        },
        "signature": "0x8d9e86ac47704f4dc6be13a79fbf99d4bab22506794d831b86ede1bebd1b23b92c7717442cb96e9fb698868d2c77521009fd469fd4445d87f6a0976a7187fe86ef6cf606c3d87461d8e45bb9943c7a3cae305973f24fe74d9b3945c5d8165a27"
      }
    }
  }
]```

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (jimmygchen) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: This check has already been performed when obtaining the `block_root`, and it returns a rejection (`404`) if block does not exist:

https://github.com/sigp/lighthouse/blob/f1f04bc68a5506c839693d25bc8c626c48ec8669/beacon_node/http_api/src/block_id.rs#L82-L92

https://github.com/sigp/lighthouse/blob/f1f04bc68a5506c839693d25bc8c626c48ec8669/beacon_node/http_api/src/block_id.rs#L116-L120

So I believe we could just supply `BlobSidecarList::default()` as default if `get_blobs` returns `None`, and even remove the `Option` from method signature and just return `Result<BlobSidecarList<T::EthSpec>, Error>`?

**Review by jimmygchen — CHANGES_REQUESTED** [review]
@qu0b thanks for submitting this PR! 🙏  

I think you're right - we should indeed return empty arrays if there are no blobs instead of returning an error. I only have small comment regarding the block lookup, I think that's already being handled.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (qu0b) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Thanks for the feedback! I did not see that check! I've implemented what you recommended

**Comment by qu0b** [comment]
Thanks for the feedback @jimmygchen hope its better now!

**Comment by jimmygchen** [comment]
Thanks for address the feedback @qu0b!

I didn't realize this function lives in `beacon_chain` crate and is used by the [p2p interface](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md) - that was an oversight from my part, should've probably suggested doing the logic the [`blob_sidecar_list`](https://github.com/sigp/lighthouse/blob/69ff71d92e22dec2ac14c5dad14272bd4e7015a8/beacon_node/http_api/src/block_id.rs#L256-L264) function in the `http_api` crate instead for an HTTP api specific change.

But now looking at the changes I think it might actually be a good cleanup, the existing debug logs we had to handle the `None` cases seems redundant, unless I'm missing something. It would be great to get a second pair of eyes from @realbigsean on this :pray:.

**Comment by realbigsean** [comment]
Looks good! Only thing is it looks like we won't return a 404 on a missing block, so we could add something like thisL

```
if blob_sidecar_list_filtered.is_empty() {
    let _ = block_id.blinded_block(&chain)?;
}
```

**Comment by jimmygchen** [comment]
@realbigsean the check was actually there before but I suggested removing, because I thought we the block root lookup in `chain.root` would cover this?

See comment here
https://github.com/sigp/lighthouse/pull/4535#discussion_r1272284846

**Comment by realbigsean** [comment]
Oh nice! My bad, I missed that comment. This looks good to me then! I think there's just a couple CI issues

**Comment by jimmygchen** [comment]
Thanks @realbigsean !

@qu0b would you mind merging latest ~~`unstable`~~ `deneb-free-blobs` into your branch and fix the CI issues. You can use the following commands: 
- Fix formatting with `cargo fmt --all` 
- Fix simple clippy errors with `make lint-fix`, you might need to fix some errors manually

Once that's fixed, I think we're good to merge! 🚀

**Comment by qu0b** [comment]
@jimmygchen My PR is against [sigp:deneb-free-blobs](https://github.com/sigp/lighthouse/tree/deneb-free-blobs) and not against unstable. Are you sure I should merge in unstable? I rebased against deneb-free-blobs and pushed those changes. I also ran `make lint-fix` and `make cargo-fmt`. Could you please approve the workflow to run again, hope it goes through this time.

**Comment by jimmygchen** [comment]
Yep, you're right  :relaxed: My apologies! 
Looks like there are some weird failures not related to your changes, I'm going to try again just in case.

**Comment by jimmygchen** [comment]
It looks like there are two failed jobs that are due to compilation errors in test code which are disabled in debug (`#![cfg(not(debug_assertions))]`), so compiler doesn't catch them unless you're running in non-debug mode. Seems like all are related to the `Option` signature change and require update in test code:
- `beacon-chain-tests` (https://github.com/sigp/lighthouse/actions/runs/5667017131/job/15360246630?pr=4535)
- `cargo-udeps` (same 6 errors as above)

Please ignore the other failures (`release-tests-windows` & `debug-tests-ubuntu`), I don't think they're related to this PR. 

Thanks for your patience 🙏  I wasn't expecting that small suggestion to trigger so many changes.

**Comment by qu0b** [comment]
@jimmygchen I will have a look and update the branch with some fixes for the failed CI steps. Thanks for the information on how to debug them.

**Comment by qu0b** [comment]
@jimmygchen I've pushed a fix, hope it works this time 🙏

**Review by jimmygchen — APPROVED** [review]
Nice 🎉  Thank you!

**Comment by qu0b** [comment]
Awesome, special thanks to [antithesis](https://antithesis.com/)

**Pull request merged**
